In [3]:
import pandas as pd
import nltk

data_1 = pd.read_csv('C://Users//Lyndon Quadros//Documents//ML Studies//Fake News Analysis//fake-news//fake.csv')

In [4]:
data_1.isnull().sum()

uuid                     0
ord_in_thread            0
author                2424
published                0
title                  680
text                    46
language                 0
crawled                  0
site_url                 0
country                176
domain_rank           4223
thread_title            12
spam_score               0
main_img_url          3643
replies_count            0
participants_count       0
likes                    0
comments                 0
shares                   0
type                     0
dtype: int64

Some observations based on the above output:

1. The author column has many NULL values and if you look at the values in the csv, a lot of them have non-sensical names like 1947, # 1 NWO Hatr etc. Although it does feel like authors would be important to the type of news being published, I think that unfortunately, we cannot use the author data in this file because it doesn't seem to have any sensible data. So I will be dropping that column from the data frame.

2. The title column has 680 Nulls. There are 2 ways that I can think about how we can impute these missing values. One way is to just replace the blank values with 'untitled' and another way would be to get a gist of the text and 'create' a title. But that would need a whole other layer. So for now, I will just impute the values with 'Untitled'.

3. The text column only has 46 missing values. But if you filter the text in the csv, then you get 200+ missing values. There's something wrong with the way the csv is formatted. If you look at rows 581, 582 etc. you'll see that the values under the 'uuid', 'author', 'ord_in_thread' columns are unintuitive. So it looks like we might have to delete these. And the best way to do this would be to delete rows that do not have a numerical value in 'ord_in_thread'.

4. For the missing values for country, we can just impute with 'nocountry'

5. For the Domain Rank, we could just add either '0' or the median value. I don't think it will matter since the dataset info on kaggle just says 'data from webhoseapi.io'.

6. The missing Thread titles can be imputed with 'untitled' and the missing Image URLs can be imputed with 'nourl'

In [5]:
#Copying data_1 a new data frame data_2 and dropping the author column
data_2 = data_1.drop(columns = ['author'])

In [8]:
#Replacing missing titles with 'Untitled'
data_2.iloc[:,3].fillna('Untitled', inplace = True)

In [ ]:
#Finding out how many rows have non-numerical values in the ord_in_thread
import numpy as np
data_2.iloc[:,1].apply(lambda x: int(9999) if ~np.isreal(x) else x)

In [18]:
data_2.loc[data_2['ord_in_thread'] == '9999']

Empty DataFrame
Columns: [uuid, ord_in_thread, published, title, text, language, crawled, site_url, country, domain_rank, thread_title, spam_score, main_img_url, replies_count, participants_count, likes, comments, shares, type]
Index: []

Looks like it has no non-numerical values in the ord_in_thread column. 

In [ ]:
#Replacing missing text with 'notext'
data_2.iloc[:,4].fillna('notext', inplace = True)

#Replacing missing country with nocountry
data_2.iloc[:,8].fillna('nocountry', inplace = True)

#Replacing missing domain_rank with the median domain rank
data_2.iloc[:,9].fillna(data_2.iloc[:,9].median(), inplace = True)

#Replacing missing thread_titles with 'nothreadtitle'
data_2.iloc[:,10].fillna('nothreadtitle', inplace = True)

#Replacing missing main image url with 'noimageurl'
data_2.iloc[:,12].fillna('noimageurl', inplace = True)


In [21]:
data_2.iloc[:,10].fillna('nothreadtitle', inplace = True)


In [22]:
data_2.isnull().sum()

uuid                  0
ord_in_thread         0
published             0
title                 0
text                  0
language              0
crawled               0
site_url              0
country               0
domain_rank           0
thread_title          0
spam_score            0
main_img_url          0
replies_count         0
participants_count    0
likes                 0
comments              0
shares                0
type                  0
dtype: int64

So there are no nulls and we have dropped the author column so far